transport/httptransport/http_transport.go

```diff
@@ -0,0 +1,237 @@
+package httptransport
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+
+	"github.com/filecoin-project/boost/transport/types/transferstatus"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/libp2p/go-eventbus"
+	"github.com/libp2p/go-libp2p-core/event"
+
+	"github.com/filecoin-project/boost/transport/httptransport/pb"
+	"github.com/filecoin-project/boost/transport/types"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/filecoin-project/boost/transport"
+)
+
+var log = logging.Logger("http-transport")
+
+const (
+	readBufferSize = 4096
+)
+
+var _ transport.Transport = (*httpTransport)(nil)
+
+type httpTransport struct {
+	// TODO Construct and inject an http client here ?
+}
+
+func (h *httpTransport) Execute(ctx context.Context, transportInfo []byte, dealInfo *types.TransportDealInfo) (th *transport.TransportHandler, isComplete bool, err error) {
+	tctx, cancel := context.WithCancel(ctx)
+
+	// de-serialize transport opaque token
+	tInfo := &pb.HttpReq{}
+	if err := proto.Unmarshal(transportInfo, tInfo); err != nil {
+		return nil, false, fmt.Errorf("failed to de-serialize transport info bytes: %w", err)
+	}
+	// check that the outputFile exists
+	fi, err := os.Stat(dealInfo.OutputFile)
+	if err != nil {
+		return nil, false, fmt.Errorf("output file state error: %w", err)
+	}
+	// validate req
+	if _, err := url.Parse(tInfo.URL); err != nil {
+		return nil, false, fmt.Errorf("failed to parse url: %w", err)
+	}
+	// do we have more bytes than required already ?
+	fileSize := fi.Size()
+	if fileSize > dealInfo.DealSize {
+		return nil, false, fmt.Errorf("deal size=%d but file size=%d", dealInfo.DealSize, fileSize)
+	}
+
+	// is the transfer already complete ? we check this by comparing the number of bytes
+	// in the output file with the deal size.
+	if fileSize == dealInfo.DealSize {
+		return nil, true, nil
+	}
+
+	pub, sub, err := createPubSub(eventbus.NewBus())
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to create pub-sub: %w", err)
+	}
+	// emit event with initial state
+	if err := pub.Emit(types.TransportEvent{
+		Status:         transferstatus.Initiated,
+		NBytesReceived: fileSize,
+	}); err != nil {
+		log.Errorw("failed to publish transport event", "dealUuid", dealInfo.DealUuid, "err", err)
+	}
+
+	var wg sync.WaitGroup
+	// start executing the transfer
+	t := &transfer{
+		tInfo:          tInfo,
+		dealInfo:       dealInfo,
+		pub:            pub,
+		nBytesReceived: fileSize,
+		wg:             wg,
+	}
+
+	t.wg.Add(1)
+	go t.execute(tctx)
+	return transport.NewHandler(tctx, cancel, dealInfo, sub, wg), false, nil
+}
+
+type transfer struct {
```

Comment:
transfer and TransportHandler share several of the same fields - any reason not to unite them into one struct?